File: app/src/main/java/org/dhis2/usescases/teiDashboard/DashboardRepository.java
package org.dhis2.usescases.teiDashboard;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import org.dhis2.commons.data.tuples.Pair;
import org.dhis2.commons.data.tuples.Trio;
import org.hisp.dhis.android.core.category.CategoryCombo;
import org.hisp.dhis.android.core.category.CategoryOptionCombo;
import org.hisp.dhis.android.core.enrollment.Enrollment;
import org.hisp.dhis.android.core.enrollment.EnrollmentStatus;
import org.hisp.dhis.android.core.event.Event;
import org.hisp.dhis.android.core.event.EventStatus;
import org.hisp.dhis.android.core.organisationunit.OrganisationUnit;
import org.hisp.dhis.android.core.program.Program;
import org.hisp.dhis.android.core.program.ProgramIndicator;
import org.hisp.dhis.android.core.program.ProgramStage;
import org.hisp.dhis.android.core.program.ProgramTrackedEntityAttribute;
import org.hisp.dhis.android.core.relationship.RelationshipType;
import org.hisp.dhis.android.core.trackedentity.TrackedEntityAttributeValue;
import org.hisp.dhis.android.core.trackedentity.TrackedEntityInstance;

import java.util.List;

import io.reactivex.Flowable;
import io.reactivex.Observable;
import io.reactivex.Single;

public interface DashboardRepository {

    Observable<List<ProgramStage>> getProgramStages(String programStages);

    Observable<Enrollment> getEnrollment();

    Observable<List<Event>> getTEIEnrollmentEvents(String programUid, String teiUid);

    Observable<List<Event>> getEnrollmentEventsWithDisplay(String programUid, String teiUid);

    Observable<List<TrackedEntityAttributeValue>> getTEIAttributeValues(String programUid, String teiUid);

    Flowable<List<ProgramIndicator>> getIndicators(String programUid);

    boolean setFollowUp(String enrollmentUid);

    Event updateState(Event event, EventStatus newStatus);

    Flowable<Enrollment> completeEnrollment(@NonNull String enrollmentUid);

    Observable<ProgramStage> displayGenerateEvent(String eventUid);

    Observable<Trio<ProgramIndicator, String, String>> getLegendColorForIndicator(ProgramIndicator programIndicator, String value);

    Integer getObjectStyle(String uid);

    Observable<List<Pair<RelationshipType, String>>> relationshipsForTeiType(String teType);

    Observable<CategoryCombo> catComboForProgram(String program);

    boolean isStageFromProgram(String stageUid);

    Observable<List<CategoryOptionCombo>> catOptionCombos(String catComboUid);

    void setDefaultCatOptCombToEvent(String eventUid);

    // FROM METADATA REPOSITORY
    Observable<TrackedEntityInstance> getTrackedEntityInstance(String teiUid);

    Observable<List<ProgramTrackedEntityAttribute>> getProgramTrackedEntityAttributes(String programUid);

    Observable<List<OrganisationUnit>> getTeiOrgUnits(@NonNull String teiUid, @Nullable String programUid);

    Observable<List<Program>> getTeiActivePrograms(String teiUid, boolean showOnlyActive);

    Observable<List<Enrollment>> getTEIEnrollments(String teiUid);

    void saveCatOption(String eventUid, String catOptionComboUid);

    Single<Boolean> deleteTeiIfPossible();

    Single<Boolean> deleteEnrollmentIfPossible(String enrollmentUid);

    Single<Integer> getNoteCount();

    EnrollmentStatus getEnrollmentStatus(String enrollmentUid);

    Observable<StatusChangeResultCode> updateEnrollmentStatus(String enrollmentUid, EnrollmentStatus status);

    boolean programHasRelationships();

    boolean programHasAnalytics();

    String getTETypeName();
}

File: app/src/main/java/org/dhis2/usescases/teiDashboard/DashboardRepository.kt
package org.dhis2.usescases.teiDashboard

import io.reactivex.Flowable
import io.reactivex.Observable
import io.reactivex.Single
import org.dhis2.commons.data.tuples.Pair
import org.dhis2.ui.MetadataIconData
import org.hisp.dhis.android.core.category.CategoryOptionCombo
import org.hisp.dhis.android.core.enrollment.Enrollment
import org.hisp.dhis.android.core.enrollment.EnrollmentStatus
import org.hisp.dhis.android.core.event.Event
import org.hisp.dhis.android.core.event.EventStatus
import org.hisp.dhis.android.core.organisationunit.OrganisationUnit
import org.hisp.dhis.android.core.program.Program
import org.hisp.dhis.android.core.program.ProgramStage
import org.hisp.dhis.android.core.program.ProgramTrackedEntityAttribute
import org.hisp.dhis.android.core.relationship.RelationshipType
import org.hisp.dhis.android.core.trackedentity.TrackedEntityAttribute
import org.hisp.dhis.android.core.trackedentity.TrackedEntityAttributeValue
import org.hisp.dhis.android.core.trackedentity.TrackedEntityInstance

interface DashboardRepository {

    fun getTeiHeader(): String?

    fun getTeiProfilePath(): String?

    fun getProgramStages(programStages: String): Observable<List<ProgramStage>>

    fun getEnrollment(): Observable<Enrollment>

    fun getTEIEnrollmentEvents(programUid: String?, teiUid: String): Observable<List<Event>>

    fun getEnrollmentEventsWithDisplay(
        programUid: String?,
        teiUid: String,
    ): Observable<List<Event>>

    fun getTEIAttributeValues(
        programUid: String?,
        teiUid: String,
    ): Observable<List<TrackedEntityAttributeValue>>

    fun setFollowUp(enrollmentUid: String?): Boolean

    fun updateState(event: Event?, newStatus: EventStatus): Event

    fun completeEnrollment(enrollmentUid: String): Flowable<Enrollment>

    fun displayGenerateEvent(eventUid: String?): Observable<ProgramStage>

    fun relationshipsForTeiType(
        teType: String,
    ): Observable<List<Pair<RelationshipType?, String>>>

    fun catOptionCombo(catComboUid: String?): CategoryOptionCombo

    fun getTrackedEntityInstance(
        teiUid: String,
    ): Observable<TrackedEntityInstance>

    fun getProgramTrackedEntityAttributes(
        programUid: String?,
    ): Observable<List<ProgramTrackedEntityAttribute>>

    fun getTeiOrgUnits(
        teiUid: String,
        programUid: String?,
    ): Observable<List<OrganisationUnit>>

    fun getTeiActivePrograms(
        teiUid: String,
        showOnlyActive: Boolean,
    ): Observable<List<kotlin.Pair<Program, MetadataIconData>>>

    fun getTEIEnrollments(
        teiUid: String,
    ): Observable<List<Enrollment>>

    fun saveCatOption(eventUid: String?, catOptionComboUid: String?)

    fun checkIfDeleteTeiIsPossible(): Boolean

    fun deleteTei(): Single<Boolean>

    fun checkIfDeleteEnrollmentIsPossible(enrollmentUid: String): Boolean

    fun deleteEnrollment(enrollmentUid: String): Single<Boolean>

    fun getNoteCount(): Single<Int>

    fun getEnrollmentStatus(enrollmentUid: String?): EnrollmentStatus?

    fun updateEnrollmentStatus(
        enrollmentUid: String,
        status: EnrollmentStatus,
    ): Observable<StatusChangeResultCode>

    fun programHasRelationships(): Boolean

    fun programHasAnalytics(): Boolean

    fun getTETypeName(): String?

    fun getAttributesMap(
        programUid: String,
        teiUid: String,
    ): Observable<List<kotlin.Pair<TrackedEntityAttribute, TrackedEntityAttributeValue>>>

    fun getDashboardModel(): DashboardModel

    fun getGrouping(): Boolean

    fun setGrouping(groupEvent: Boolean)
}
